### PR TITLE
Make secio almost compile for asmjs/wasm

### LINKED ISSUE #499: Make secio compile for wasm/emscripten
The problematic crates are `ring` and `secp256k1`.

There are four major components in secio:
- Public key verification. We could use `ed25519-dalek`. I'm not aware of any crate apart from `ring` that supports RSA, but we could disable it conditionally, like secp.
- The digests. We could use the `sha2` crate, or similar.
- The ciphers. This should already work.
- The agreement algorithm (ECDH). This is problematic, as I don't know any wasm/emscripten-friendly crate that does that.

For the agreement, we could wrap platform-specific code that uses WebCrypto underneath. Since it is only used in the handshake, it should be possible to adjust for the asynchronicity of WebCrypto.

### PR DESCRIPTION
**Pull request by tomaka**
Close #499 

~~Disclaimer: this doesn't actually compile because emscripten doesn't support i128 operations. However on the Rust side everything works, and one can argue that this PR is a good first step. I'm also fine with leaving this PR on ice for some time.~~
For some reason this works now.

Disclaimer: the crypto library I used (`ed25519-dalek`) is not audited, but neither is ring.

This PR replaces most of the usages of `ring` with the `ed25519-dalek`, `hmac` and `sha2` crates. The only remaining usage is for ECDH, because I'm not aware of any other library that supports this. For emscripten, the implementation uses `WebCrypto`. The `WebCrypto` path is unfortunately untested, as the code doesn't actually compile.

This adds a `rsa` feature to secio which makes it support RSA public keys. If we want to connect to IPFS nodes, this feature has to be enabled.
I also removed the `libp2p-secio` feature, making secio mandatory.

### CODE REVIEW

**Comment by tomaka** [comment]
Successfully tested on the polkadot testnet.

**Comment by tomaka** [comment]
Compiling a binary still doesn't work because of i128s not being supported.

File: protocols/secio/src/algo_support.rs [hunk 1]
```diff
@@ -77,16 +71,16 @@ where
 ///
 /// The `Ordering` parameter determines which argument is preferred. If `Less` or `Equal` we
 /// try for each of `theirs` every one of `ours`, for `Greater` it's the other way around.
-pub fn select_agreement<'a>(r: Ordering, ours: &str, theirs: &str) -> Result<&'a agreement::Algorithm, SecioError> {
+pub fn select_agreement<'a>(r: Ordering, ours: &str, theirs: &str) -> Result<KeyAgreement, SecioError> {
```
Comment (twittner) [inline protocols/secio/src/algo_support.rs]: `<'a>` can be removed.

File: protocols/secio/src/algo_support.rs [hunk 2]
```diff
@@ -182,20 +187,30 @@ where
 ///
 /// The `Ordering` parameter determines which argument is preferred. If `Less` or `Equal` we
 /// try for each of `theirs` every one of `ours`, for `Greater` it's the other way around.
-pub fn select_digest<'a>(r: Ordering, ours: &str, theirs: &str) -> Result<&'a digest::Algorithm, SecioError> {
+pub fn select_digest<'a>(r: Ordering, ours: &str, theirs: &str) -> Result<Digest, SecioError> {
```
Comment (twittner) [inline protocols/secio/src/algo_support.rs]: `<'a>` can be removed.

File: protocols/secio/src/exchange/impl_ring.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+/// Opaque private key type.
+pub type AgreementPrivateKey = ring_agreement::EphemeralPrivateKey;
+
+/// Generates a new key pair as part of the exchange.
+///
+/// Returns the opaque private key and the corresponding public key.
+pub fn generate_agreement(algorithm: KeyAgreement) -> impl Future<Item = (AgreementPrivateKey, Vec<u8>), Error = SecioError> {
+    let rng = ring_rand::SystemRandom::new(); // TODO: don't recreate that here?
```
Comment (twittner) [inline protocols/secio/src/exchange/impl_ring.rs]: [Should be fine.](https://docs.rs/ring/0.13.2/src/ring/rand.rs.html#87)

File: protocols/secio/src/codec/mod.rs [hunk 4]
```diff
@@ -37,6 +39,61 @@ pub type FullCodec<S> = DecoderMiddleware<EncoderMiddleware<length_delimited::Fr
...
+            Hmac::Sha512(_) => 64,
+        }
+    }
+
+    /// Builds a `Hmac` from an algorithm and key.
+    pub fn from_key(algorithm: Digest, key: &[u8]) -> Self {
+        // TODO: don't unwrap
+        match algorithm {
+            Digest::Sha256 => Hmac::Sha256(Mac::new_varkey(key).unwrap()),
+            Digest::Sha512 => Hmac::Sha512(Mac::new_varkey(key).unwrap()),
```
Comment (twittner) [inline protocols/secio/src/codec/mod.rs]: Maybe replace `unwrap` with `expect` explaining that `HMAC` accepts keys of any size.

**Pull request merged**